java-spiffe-core/src/main/java/spiffe/internal/CertificateUtils.java

```diff
@@ -0,0 +1,269 @@
+     * defined in SAN_VALUE_INDEX and filters the entries that starts with the SPIFFE_PREFIX and returns the first.
+     *
+     * @param certificate a {@link X509Certificate}
+     * @return an instance of a {@link SpiffeId}
+     * @throws CertificateException if the certificate contains multiple SPIFFE IDs, or does not contain any, or
+     *                              the SAN extension cannot be decoded
+     */
+    public static SpiffeId getSpiffeId(X509Certificate certificate) throws CertificateException {
+        val spiffeIds = getSpiffeIds(certificate);
+
+        if (spiffeIds.size() > 1) {
+            throw new CertificateException("Certificate contains multiple SPIFFE IDs");
+        }
+
+        if (spiffeIds.size() < 1) {
+            throw new CertificateException("Certificate does not contain SPIFFE ID in the URI SAN");
+        }
+
+        return SpiffeId.parse(spiffeIds.get(0));
+    }
+
+    /**
+     * Extracts the trust domain of a chain of certificates.
+     *
+     * @param chain a list of {@link X509Certificate}
+     * @return a {@link TrustDomain}
+     * @throws CertificateException
+     */
+    public static TrustDomain getTrustDomain(List<X509Certificate> chain) throws CertificateException {
+        val spiffeId = getSpiffeId(chain.get(0));
+        return spiffeId.getTrustDomain();
+    }
+
+    /**
+     * Validates that the private key and the public key in the x509Certificate match by
+     * creating a signature with the private key and verifying with the public key.
+     *
+     * @throws InvalidKeyException if the keys don't match
+     */
+    public static void validatePrivateKey(PrivateKey privateKey, X509Certificate x509Certificate) throws InvalidKeyException {
+        // create a challenge
+        byte[] challenge = new byte[1000];
+        ThreadLocalRandom.current().nextBytes(challenge);
+
+        Signature sig = null;
+
+        try {
+            Algorithm.Family algorithm = Algorithm.Family.parse(privateKey.getAlgorithm());
+            if (Algorithm.Family.RSA.equals(algorithm)) {
+                sig = Signature.getInstance("SHA256withRSA");
+            } else if (Algorithm.Family.EC.equals(algorithm)) {
+                sig = Signature.getInstance("SHA1withECDSA");
+            } else {
+                throw new InvalidKeyException(String.format("Private Key algorithm not supported: %s", algorithm));
+            }
+
+            sig.initSign(privateKey);
+            sig.update(challenge);
+            byte[] signature = sig.sign();
+
+            sig.initVerify(x509Certificate.getPublicKey());
+            sig.update(challenge);
+
+            if (!sig.verify(signature)) {
+                throw new InvalidKeyException("Private Key does not match Certificate Public Key");
+            }
+        } catch (SignatureException | NoSuchAlgorithmException e) {
+            throw new IllegalStateException("Could not validate private keys", e);
+        }
+    }
```

Comment:
As the `X509Svid` allows to parse SVIDs from bytes or from files, it could be potentially used to parse certificates from non-trusted sources, we need to validate that the private and public keys match, and some flags in the certificate are validated as well, to ensure that the X509Svid created is valid. All these validations are consistent with what was implemented in the  `go-spiffe` library.
I improved the validation for generating the challenge in a secure fashion and using non-compromised algorithms. (Btw, validating only the ECParameterSpec doesn't ensure that a given private key pairs with a given public key, it's not sufficient.)